File: LeetCode/Patterns/Stack & Queue/496. Next Greater Element I/Solution.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        mp = {}
        st = []

        for num in nums2:
            while st and st[-1] < num:
                mp[st.pop()] = num
            st.append(num)

        res = []
        for num in nums1:
            if num in mp:
                res.append(mp[num])
            else:
                res.append(-1)

        return res

File: LeetCode/Patterns/Stack & Queue/496. Next Greater Element I/Solution.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        int n = nums1.size();
        int m = nums2.size();

        unordered_map<int, int> map;
        vector<int> res;
        stack<int> st;
        
        for(int num : nums2) {
            while(!st.empty() && num > st.top()) {
                map[st.top()] = num;
                st.pop();
            }
            st.push(num);
        }

        for(int num : nums1) {
            if(map.count(num))
                res.push_back(map[num]);
            else 
                res.push_back(-1);
        }
        return res;
    }
};